node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+	/// Includes the `K` ancestors too.
+	/// Maps relay parent / ancestor -> live candidate receipts.
+	receipts: HashMap<H256, HashSet<CommittedCandidateReceipt>>,
+
+	/// Allow reverse caching of view checks.
+	/// Maps candidate hash -> relay parent for extracting meta information from `PerRelayParent`.
+	/// Note that the presence of this is not sufficient to determine if deletion is ok, i.e.
+	/// two histories could cover this.
+	reverse: HashMap<H256, H256>,
+
+	/// Keeps track of which candidate receipts are required due to ancestors of which relay parents
+	/// of our view.
+	/// Maps ancestor -> relay parents in view
+	ancestry: HashMap<H256, HashSet<H256>>,
+
+	/// Track things needed to start and stop work on a particular relay parent.
+	per_relay_parent: HashMap<H256, PerRelayParent>,
+
+	/// Track data that is specific to a candidate.
+	per_candidate: HashMap<H256, PerCandidate>,
+}
+
+#[derive(Debug, Clone, Default)]
+struct PerCandidate {
+	/// A Candidate and a set of known erasure chunks in form of messages to be gossiped / distributed if the peer view wants that.
+	/// This is _across_ peers and not specific to a particular one.
+	/// candidate hash + erasure chunk index -> gossip message
+	message_vault: HashMap<u32, AvailabilityGossipMessage>,
+
+	/// Track received candidate hashes and chunk indices from peers.
+	received_messages: HashMap<PeerId, HashSet<(H256, ValidatorIndex)>>,
+
+	/// Track already sent candidate hashes and the erasure chunk index to the peers.
+	sent_messages: HashMap<PeerId, HashSet<(H256, ValidatorIndex)>>,
+}
+
+#[derive(Debug, Clone, Default)]
+struct PerRelayParent {
+	/// Set of `K` ancestors for this relay parent.
+	ancestors: HashSet<H256>,
+
+	/// The set of validators.
+	validators: Vec<ValidatorId>,
+
+	/// If this node is a validator, note the index in the validator set.
+	validator_index: Option<ValidatorIndex>,
+}
+
+impl ProtocolState {
+	/// Unionize all cached entries for the given relay parents and it's ancestors ancestors.
+	/// Ignores all non existant relay parents, so this can be used directly with a peers view.
+	/// Returns a map from candidate hash -> receipt
+	fn cached_live_candidates_unioned<'a>(
+		&'a self,
+		relay_parents: impl IntoIterator<Item = &'a H256> + 'a,
+	) -> HashMap<H256, CommittedCandidateReceipt> {
+		let relay_parents_and_ancestors = relay_parents
+			.into_iter()
+			.map(|relay_parent| {
+				self
+					.per_relay_parent
+					.get(relay_parent)
+					.into_iter()
+					.map(|per_relay_parent| {
+						per_relay_parent.ancestors.iter().cloned()
+					})
+					.flatten()
+					.chain(iter::once(relay_parent.clone()))
+			})
+			.flatten()
+			.collect::<HashSet<H256>>();
+
+		relay_parents_and_ancestors
+			.into_iter()
+			.filter_map(|relay_parent_or_ancestor| self.receipts.get(&relay_parent_or_ancestor))
+			.map(|receipt_set| receipt_set.into_iter())
+			.flatten()
+			.map(|receipt| (candidate_hash_of(receipt), receipt.clone()))
```

Comment:
maybe better for the stored data in `self.receipts` to cache this hash, so we don't do hundreds of hashes on every relay-chain block

Reply:
Good point, the assumption that the `Hasher` used by `HashMap` (`hashbrown`, or are we using something else/planning on using smth else?) is significantly faster than `H256` is always true :+1:

Reply:
Yeah, blake2-256 is cryptographically secure whereas AHash is meant to be fast.

However, we should also have a utility at some point where we create a `HashMap` type alias using an identity hashes. We don't need do any hashing of the keys, because the keys are already hashes with more than sufficient entropy. But that is a premature optimization.